File: src/main/java/com/muxiu1997/sharewhereiam/network/MessageShareWaypoint.java
package com.muxiu1997.sharewhereiam.network;

import com.muxiu1997.sharewhereiam.util.ShareWaypointUtil;
import cpw.mods.fml.common.network.ByteBufUtils;
import cpw.mods.fml.common.network.simpleimpl.IMessage;
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
import cpw.mods.fml.common.network.simpleimpl.MessageContext;
import io.netty.buffer.ByteBuf;
import journeymap.client.model.Waypoint;

import javax.annotation.Nullable;

public class MessageShareWaypoint implements IMessage {

    public String playerName;
    public Waypoint waypoint;
    public String additionalInformation;

    public MessageShareWaypoint() {
    }

    public MessageShareWaypoint(String playerName, Waypoint waypoint, String additionalInformation) {
        this.playerName = playerName;
        this.waypoint = waypoint;
        this.additionalInformation = additionalInformation;
    }

    @Override
    public void fromBytes(ByteBuf buf) {
        playerName = ByteBufUtils.readUTF8String(buf);
        waypoint = Waypoint.fromString(ByteBufUtils.readUTF8String(buf));
        additionalInformation = ByteBufUtils.readUTF8String(buf);
    }

    @Override
    public void toBytes(ByteBuf buf) {
        ByteBufUtils.writeUTF8String(buf, playerName);
        ByteBufUtils.writeUTF8String(buf, waypoint.toString());
        ByteBufUtils.writeUTF8String(buf, additionalInformation);
    }

    public static class Handler implements IMessageHandler<MessageShareWaypoint, IMessage> {
        @Override
        @Nullable
        public IMessage onMessage(MessageShareWaypoint message, MessageContext ctx) {
            if (ctx.side.isServer()) return null;
            ShareWaypointUtil.addShareWaypointChat(message.playerName, message.waypoint, message.additionalInformation);
            return null;
        }
    }
}

File: src/main/kotlin/com/muxiu1997/sharewhereiam/network/MessageShareWaypoint.kt
package com.muxiu1997.sharewhereiam.network

import com.muxiu1997.sharewhereiam.util.WaypointUtil
import cpw.mods.fml.common.network.ByteBufUtils
import cpw.mods.fml.common.network.simpleimpl.IMessage
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler
import cpw.mods.fml.common.network.simpleimpl.MessageContext
import cpw.mods.fml.relauncher.Side
import cpw.mods.fml.relauncher.SideOnly
import io.netty.buffer.ByteBuf
import journeymap.client.cartography.RGB
import journeymap.client.model.Waypoint
import net.minecraft.client.Minecraft

class MessageShareWaypoint : IMessage {
    lateinit var playerName: String
    lateinit var waypointJson: String
    lateinit var additionalInformation: String

    @Suppress("unused")
    constructor()

    @SideOnly(Side.CLIENT)
    constructor(playerWaypoint: WaypointUtil.PlayerWaypoint) : this(
        playerWaypoint,
        ""
    )

    @SideOnly(Side.CLIENT)
    constructor(playerWaypoint: WaypointUtil.PlayerWaypoint, additionalInformation: String) {
        this.playerName = playerWaypoint.player.displayName
        this.waypointJson = playerWaypoint.waypoint.toString()
        this.additionalInformation = additionalInformation
    }

    override fun fromBytes(buf: ByteBuf) {
        playerName = ByteBufUtils.readUTF8String(buf)
        waypointJson = ByteBufUtils.readUTF8String(buf)
        additionalInformation = ByteBufUtils.readUTF8String(buf)
    }

    override fun toBytes(buf: ByteBuf) {
        ByteBufUtils.writeUTF8String(buf, playerName)
        ByteBufUtils.writeUTF8String(buf, waypointJson)
        ByteBufUtils.writeUTF8String(buf, additionalInformation)
    }

    companion object {
        object Handler : IMessageHandler<MessageShareWaypoint, IMessage?>, IClientSideHandler, IServerSideHandler {
            override fun onMessage(message: MessageShareWaypoint, ctx: MessageContext): IMessage? {
                if (ctx.side.isClient) {
                    handleClientSideMessage(message)
                } else {
                    handleServerSideMessage(message)
                }
                return null
            }

            @SideOnly(Side.CLIENT)
            private fun handleClientSideMessage(message: MessageShareWaypoint) {
                WaypointUtil.addShareWaypointChat(
                    message.playerName,
                    Waypoint.fromString(message.waypointJson),
                    message.additionalInformation
                )
            }

            private fun handleServerSideMessage(message: MessageShareWaypoint) {
                network.sendToAll(message)
            }
        }
    }
}
